src/Paprika/Utils/WorkProcessor.cs

```diff
@@ -0,0 +1,181 @@
+    {
+        _data = data;
+    }
+
+    public void Execute()
+    {
+        var i = _data.Index.GetNext();
+        while (i < _data.ToExclusive)
+        {
+            _data.Action(i);
+            i = _data.Index.GetNext();
+        }
+
+        _data.MarkThreadCompleted();
+    }
+
+    private class SharedCounter(int fromInclusive)
+    {
+        private int _index = fromInclusive;
+        public int GetNext() => Interlocked.Increment(ref _index) - 1;
+    }
+
+    private class Data(int threads, int fromInclusive, int toExclusive, Action<int> action)
+    {
+        public SharedCounter Index { get; } = new SharedCounter(fromInclusive);
+        public int ToExclusive => toExclusive;
+        public Action<int> Action => action;
+        public int ActiveThreads => threads;
+
+        public int MarkThreadCompleted()
+        {
+            var remaining = Interlocked.Decrement(ref threads);
+
+            if (remaining == 0)
+            {
+                lock (this)
+                {
+                    Monitor.Pulse(this);
+                }
+            }
+
+            return remaining;
+        }
+    }
+
+    private class InitProcessor<TLocal> : IThreadPoolWorkItem
+    {
+        private readonly Data<TLocal> _data;
+
+        public static void For(
+            int fromInclusive,
+            int toExclusive,
+            ParallelOptions parallelOptions,
+            Func<TLocal> init,
+            Func<int, TLocal, TLocal> action,
+            Action<TLocal> @finally)
+        {
+            var threads = parallelOptions.MaxDegreeOfParallelism > 0 ? parallelOptions.MaxDegreeOfParallelism : Environment.ProcessorCount;
+
+            var data = new Data<TLocal>(threads, fromInclusive, toExclusive, init, action, @finally);
+
+            for (int i = 0; i < threads - 1; i++)
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(new InitProcessor<TLocal>(data), preferLocal: false);
+            }
+
+            new InitProcessor<TLocal>(data).Execute();
+
+            if (data.ActiveThreads > 0)
+            {
+                lock (data)
+                {
+                    if (data.ActiveThreads > 0)
+                    {
+                        // Wait for remaining to complete
+                        Monitor.Wait(data);
+                    }
+                }
+            }
+        }
+
+        private InitProcessor(Data<TLocal> data)
+        {
+            _data = data;
+        }
+
+        public void Execute()
+        {
+            var value = _data.Init();
+            var i = _data.Index.GetNext();
+            while (i < _data.ToExclusive)
+            {
+                value = _data.Action(i, value);
+                i = _data.Index.GetNext();
+            }
+
+            _data.Finally(value);
+
+            _data.MarkThreadCompleted();
+        }
+
+        private class Data<TValue>(int threads, int fromInclusive, int toExclusive, Func<TValue> init,
```

Comment:
Have you considered deriving from `Data` here? A lot is shared, beside the `local` related concepts?

Reply:
Moved to derived classes